src/core/meshcop/tcat_agent.cpp

```diff
@@ -617,13 +639,129 @@ Error TcatAgent::HandleGetProvisioningUrl(Message &aOutgoingMessage, bool &respo
     length = StringLength(mVendorInfo->mProvisioningUrl, kProvisioningUrlMaxLength);
     VerifyOrExit(length > 0 && length <= Tlv::kBaseTlvMaxLength, error = kErrorInvalidState);
 
-    error    = Tlv::AppendTlv(aOutgoingMessage, kTlvResponseWithPayload, mVendorInfo->mProvisioningUrl, length);
-    response = true;
+    SuccessOrExit(error =
+                      Tlv::AppendTlv(aOutgoingMessage, kTlvResponseWithPayload, mVendorInfo->mProvisioningUrl, length));
+    aResponse = true;
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandlePresentPskdHash(const Message &aIncomingMessage, uint16_t aOffset, uint16_t aLength)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(mVendorInfo->mPskdString != nullptr, error = kErrorSecurity);
+
+    error = VerifyHash(aIncomingMessage, aOffset, aLength, mVendorInfo->mPskdString,
+                       StringLength(mVendorInfo->mPskdString, kMaxPskdLength));
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandlePresentPskcHash(const Message &aIncomingMessage, uint16_t aOffset, uint16_t aLength)
+{
+    Error         error = kErrorNone;
+    Dataset::Info datasetInfo;
+    Pskc          pskc;
+
+    VerifyOrExit(Get<ActiveDatasetManager>().Read(datasetInfo) == kErrorNone, error = kErrorSecurity);
+    VerifyOrExit(datasetInfo.IsPresent<Dataset::kPskc>(), error = kErrorSecurity);
+    pskc = datasetInfo.Get<Dataset::kPskc>();
+
+    error = VerifyHash(aIncomingMessage, aOffset, aLength, pskc.m8, Pskc::kSize);
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandlePresentInstallCodeHash(const Message &aIncomingMessage, uint16_t aOffset, uint16_t aLength)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(mVendorInfo->mInstallCode != nullptr, error = kErrorSecurity);
+
+    error = VerifyHash(aIncomingMessage, aOffset, aLength, mVendorInfo->mInstallCode,
+                       StringLength(mVendorInfo->mInstallCode, 255));
 
```

Comment:
The install code challenge is currently not relevant command class authentication. Instead the application layer may need to know if the install code was presented by commissioner to elevate the rights in application context.   

I suggest storing the result for the duration of the session and adding an application API for InstallCode.
Or should this go in a future pull request?

Reply:
Here we voted (I think) for the future PR to address this. Not required for initial cert tests.